// ion2-calendar/dist/index.js
"use strict";
function __export(m) {
    for (var p in m) if (!exports.hasOwnProperty(p)) exports[p] = m[p];
}
Object.defineProperty(exports, "__esModule", { value: true });
__export(require("./calendar.model"));
var calendar_modal_1 = require("./components/calendar.modal");
exports.CalendarModal = calendar_modal_1.CalendarModal;
var calendar_week_component_1 = require("./components/calendar-week.component");
exports.CalendarWeekComponent = calendar_week_component_1.CalendarWeekComponent;
var month_component_1 = require("./components/month.component");
exports.MonthComponent = month_component_1.MonthComponent;
var calendar_options_provider_1 = require("./services/calendar-options.provider");
exports.DEFAULT_CALENDAR_OPTIONS = calendar_options_provider_1.DEFAULT_CALENDAR_OPTIONS;
var calendar_component_1 = require("./components/calendar.component");
exports.CalendarComponent = calendar_component_1.CalendarComponent;
var calendar_module_1 = require("./calendar.module");
exports.CalendarModule = calendar_module_1.CalendarModule;
var calendar_controller_1 = require("./calendar.controller");
exports.CalendarController = calendar_controller_1.CalendarController;
